c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -64,15 +64,64 @@ constexpr size_t kMinLargeAlloc = 10485760; // allocations between 1 and 10 MiB
 constexpr size_t kRoundLarge = 2097152;     // round up large allocs to 2 MiB
 
 struct DeviceStats {
-  uint64_t   amount_allocated;      // total amount allocated in bytes
-  uint64_t   max_amount_allocated;  // max total amount allocated in bytes
-  uint64_t   amount_cached;         // total amount in cache in bytes
-  uint64_t   max_amount_cached;     // max total amount in cache in bytes
+  uint64_t   total_num_alloc_requests;    // total number of allocation requests received by
+                                          // the allocator
+  uint64_t   total_num_free_requests;     // total number of free requests received by the allocator
+
+  uint64_t   total_num_blocks_allocated;  // total number of memory blocks ever created
+                                          // (including splits)
+  uint64_t   total_num_blocks_released;   // total number of memory blocks ever released
+  uint64_t   total_num_blocks_split;      // total number of memory blocks ever split into two
+
+  uint64_t   total_num_cuda_mallocs;      // total number of calls to CUDA malloc
+  uint64_t   total_num_cuda_frees;        // total number of calls to CUDA free
+  uint64_t   total_num_cache_flushes;     // total number of cache flushes
+                                          //   (i.e. failed calls to CUDA malloc)
+
+  uint64_t   amount_allocated;            // total amount allocated in bytes
+  uint64_t   max_amount_allocated;        // max total amount allocated in bytes
+  uint64_t   amount_cached;               // total amount in cache in bytes
+  uint64_t   max_amount_cached;           // max total amount in cache in bytes
 
   DeviceStats() :
+      total_num_alloc_requests(0), total_num_free_requests(0),
+      total_num_blocks_allocated(0), total_num_blocks_released(0), total_num_blocks_split(0),
+      total_num_cuda_mallocs(0), total_num_cuda_frees(0), total_num_cache_flushes(0),
       amount_allocated(0), max_amount_allocated(0),
       amount_cached(0), max_amount_cached(0) { }
 
+  void logAllocRequest(size_t occurrences=1) {
```

Comment:
I used existing convention here (see e.g. `decreaseAllocated `); I'm happy to change everything to straight `uint64_t`. re. value of these functions – also just convention from `decreaseAllocated` and friends.

My thoughts are:

- Change `size_t` to `uint64_t` across the board.
- Leave the helper methods as is (some of the helper methods do add value e.g. `increaseAllocated`, and it would be unwieldy to have a mix of plain `+=` statements and calls to helper methods.